xdsl/transforms/experimental/StencilInlining.py

```diff
@@ -0,0 +1,341 @@
+                                       ) and not isinstance(
+                                           producer_op.region.ops[
+                                               i + 1 + num_store_result_op],
+                                           ReturnOp):
+                                    access_flag = False
+                                elif i != len(
+                                        producer_op.region.ops
+                                ) - 1 and not len(producer_op_external_uses):
+                                    access_flag = False
+
+                            if not use_other_than_store_or_return and access_flag:
+                                res_final = producer_op_unit_clone_normal_op.results[
+                                    0]
+
+                                uses = list(op.results[0].uses)
+                                for use in uses:
+                                    use.operation.replace_operand(
+                                        use.index, res_final)
+
+                                uses = list(producer_op_unit.results[0].uses)
+                                for use in uses:
+                                    use.operation.replace_operand(
+                                        use.index, res_final)
+                            else:
+                                res_final = producer_op_unit_clone_normal_op.results[
+                                    0]
+
+                                uses = list(producer_op_unit.results[0].uses)
+                                for use in uses:
+                                    use.operation.replace_operand(
+                                        use.index, res_final)
+                    elif isinstance(producer_op_unit, ReturnOp):
+                        if not len(inlined_op_return_arguments):
+                            for x in list(producer_op_unit.operands):
+                                assert isinstance(x, OpResult)
+                                inlined_op_return_arguments.append(x)
+            elif not isinstance(op, ReturnOp):
+                op_clone = op.clone()
+                inlined_op_block.add_op(op_clone)
+
+                for i, res in enumerate(op.results):
+                    res_uses = list(res.uses)
+                    for use in res_uses:
+                        use.operation.replace_operand(use.index,
+                                                      op_clone.results[i])
+            else:
+                if not len(inlined_op_return_arguments):
+                    op_clone = op.clone()
+                    inlined_op_block.add_op(op_clone)
+                else:
+                    combined_list: list[SSAValue] = [
+                        *inlined_op_return_arguments, *list(op.operands)
+                    ]
+
+                    inlined_op_return = ReturnOp.get(combined_list)
+                    inlined_op_block.add_op(inlined_op_return)
```

Comment:
This specific part is still quite hard to get into from the outside. Could you maybe add high level comments as a first step?